### ISSUE TITLE
QtWebEngine backend

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#666]
# Rough TODO-list
- [x] Basic backend/browsing
- [x] Loading status
- [x] Scroll position
- [x] Settings
  - [x] Support to mark settings as QtWebKit/QtWebEngine only
  - [x] Basic QWebEngineSettings
  - [x] Proxy (blocked by #2082)
  - [x] Custom headers (`DNT`/`Accept-Language`) (?)
  - [x] User-stylesheet (relevant code in QupZilla: [1](https://github.com/QupZilla/qupzilla/blob/v2.0/src/lib/app/mainapplication.cpp#L1063-L1101), [2](https://github.com/QupZilla/qupzilla/blob/v2.0/src/lib/tools/scripts.cpp#L119-L132))
  - [x] Cookies-related settings
  - [x] Cache-related settings
  - [ ] Private browsing
  - [x] User agent
- [x] ~~Custom cookie handling~~
- [x] ~~Custom cache handling~~
- [x] Downloads
  - [x] Basic downloads / unsupported content
  - [x] Download tests
  - [x] Downloading via hints
    - [x] via QtNetwork
    - [ ] via QtWebEngine
  - [x] Downloading via `:download`
    - [x] via QtNetwork
    - [ ] via QtWebEngine
  - [x] Downloading adblock filterlists
    - [x] via QtNetwork
    - [ ] via QtWebEngine
  - [x] MHTML downloads
- [x] Authentication handling
- [x] SSL error handling
- [x] Commands
  - [x] Support to mark commands as QtWebKit/QtWebEngine only
  - [x] Make as many commands as possible work.
    - [x] `:navigate`
    - [x] `:open-editor`
    - [x] `:insert-text`
    - [x] `:paste-primary`
- [x] ~~Support for Qt 5.5+~~ (not going to happen, 5.6+ only)
- [x] Custom URL handlers (`qute://*`)
  - [ ] `qute:settings` / JS bridge
- [x] Host blocking
- [x] Hinting
  - [x] Adjust for zoom position
  - [x] Only show visible hints
  - [ ] frames
  - [ ] iframes
- [x] History
- [x] Sessions
- [x] Inspector
- [x] ~~Custom error pages~~
- [x] Feature permissions (geolocation, notifications, etc)
- [x] Fullscreen support for videos etc.
- [x] Javascript:
  - [x] log
  - [x] alert
  - [x] confirm
  - [x] open window/tab (`createWindow`)
  - [x] close window/tab
- [x] Selecting files
- [x] View source
- [x] auto-insert-mode
- [x] Mouse handling
  - [x] Insert mode on click
  - [x] Ctrl-Click / Middle-click (open links in new tab)
  - [x] Mouse wheel zooming 
  - [x] Rocker gestures
  - [x] Back/forward mouse buttons
- [x] Other customizations in `WebView`/`WebPage` subclasses
  - [x] page background color (had some issues)
  - [ ] preventing keyboard focus during insert mode (had some issues)
- [x] printing (available with Qt 5.8 - ~~`:print --pdf` is implemented but untested, normal printing is not implemented - also see https://github.com/QupZilla/qupzilla/pull/2068~~)
- [x] Make tests run for QtWebEngine
  - [x] end2end tests
    - [x] get testsuite to run cleanly
    - [x] Integrate into CI
    - [ ] Investigate why `xfail(run=False)` tests in `caret.feature` are so slow
    - [x] Add run=False to some slow xfails or reduce wait timeout
    - [ ] Investigate unknown-broken/flaky tests
  - [x] unittests
- [x] Running without QtWebKit
  - [x] Enforcing that via CI
- [ ] pdfjs? (though [QTBUG-50556](https://bugreports.qt.io/browse/QTBUG-50556) talks about adding PDFium))
- [ ] Showing audible tabs / muting

# Notes about individual features
## Hints

The hints (labels to click links via keyboard, see [screenshot](https://github.com/The-Compiler/qutebrowser/blob/master/doc/img/hints.png)) are currently implemented in Python, via the [QWebElement](http://doc.qt.io/archives/qt-5.5/qwebelement.html) API (see [hints.py](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/browser/hints.py) and [webelem.py](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/browser/webelem.py)).

Since that API doesn't exist anymore with QtWebEngine, a big part of it will have to be reimplemented in Javascript, either using [QWebEnginePage::runJavaScript](http://doc.qt.io/qt-5/qwebenginepage.html#runJavaScript) or [QWebChannel](http://doc.qt.io/qt-5/qwebchannel.html).
- [x] Write more tests for `:hint` features (see [hints.feature](https://github.com/The-Compiler/qutebrowser/blob/master/tests/integration/features/hints.feature))
- [x] Write more HTML hint tests (see [simple.html](https://github.com/The-Compiler/qutebrowser/blob/master/tests/integration/data/hints/html/simple.html))
- [x] Split current `HintManager` class (see [hints.py](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/browser/hints.py)) into separate classes for finding elements/drawing labels/clicking elements
- [x] Start implementing those parts in javascript

**Comment by addisonamiri** [comment]
I'd like this feature a lot. There's only one open source blink-based browser I know of (chromium) so you'll be doubling the number :D

**Comment by The-Compiler** [comment]
For the lack of a better place to put this - here's a great example by Kovid Goyal on how to use QtWebChannel:

```python
#!/usr/bin/env python3
# vim:fileencoding=utf-8

from PyQt5.Qt import *

qwebchannel_js = QFile(':/qtwebchannel/qwebchannel.js')
if not qwebchannel_js.open(QIODevice.ReadOnly):
    raise SystemExit(
        'Failed to load qwebchannel.js with error: %s' %
        qwebchannel_js.errorString())
qwebchannel_js = bytes(qwebchannel_js.readAll()).decode('utf-8')


def client_script():
    script = QWebEngineScript()
    script.setSourceCode(qwebchannel_js + '''
    new QWebChannel(qt.webChannelTransport, function(channel) {
        channel.objects.bridge.print('Hello world!');
    });
''')
    script.setName('xxx')
    script.setWorldId(QWebEngineScript.MainWorld)
    script.setInjectionPoint(QWebEngineScript.DocumentReady)
    script.setRunsOnSubFrames(True)
    return script


class WebPage(QWebEnginePage):

    def javaScriptConsoleMessage(self, level, msg, linenumber, source_id):
        try:
            print('%s:%s: %s' % (source_id, linenumber, msg))
        except OSError:
            pass

    @pyqtSlot(str)
    def print(self, text):
        print('From JS:', text)


app = QApplication([])
p = WebPage()
v = QWebEngineView()
v.setPage(p)
p.profile().scripts().insert(client_script())
c = QWebChannel(p)
p.setWebChannel(c)
c.registerObject('bridge', p)
v.setHtml('<p>Hello world!')
v.show()
app.exec_()
```

**Comment by vyp** [comment]
Not sure if this is the right place but I just wanted to point out https://wiki.qt.io/New_Features_in_Qt_5.6#Removed_Modules. Not sure what that means for qutebrowser though.

**Comment by The-Compiler** [comment]
@vyp See [my answer to the Archlinux BBS](https://bbs.archlinux.org/viewtopic.php?pid=1569755#p1569755) (and the comment on the Phoronix post quoted there).

**Comment by vyp** [comment]
@The-Compiler Oh I didn't know about that, sorry! Also, I'm not blaming you, just trying to point out information which may be helpful.

But I guess all this means is that as long as distributions continue to *package* qtwebkit, this shouldn't be an issue.

**Comment by The-Compiler** [comment]
No worries - I just got into a bit of rage-mode there because of the Phoronix commenter :wink:

> But I guess all this means is that as long as distributions continue to *package* qtwebkit, this shouldn't be an issue.

Indeed. And looking at how much still is using QtWebKit, I'm quite sure they'll continue packaging it rather than dropping all that.

**Comment by The-Compiler** [comment]
# Refactoring notes

Currently working on refactoring everything QtWebKit-specific - some random notes here:

- [ ] Can we simplify the `WebView.url_text_changed` signal, get rid of it, or move it to `webkittab.py`?
- [ ] How much of `webview.py` should go to `webkittab.py`?

**Comment by The-Compiler** [comment]
The `qtwebengine` branch got merged today, so now you can do `--backend webengine` to use QtWebEngine in `master` - however a lot of stuff will still be broken :wink:

**Comment by The-Compiler** [comment]
@artur-shaik Would you be interested in working on caret browsing for QtWebEngine by any chance?

Since the related web actions aren't available like in QtWebKit, you'd need to do it all in JS - you can add a new file to `javascript/` like I did with `scroll.js`, and then you'd need to implement `WebEngineCaret` in [`webenginetab.py`](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/browser/webengine/webenginetab.py).

Then you can call functions from that file similar to what I did in b78b89f04f75c3b113e62179a863dca3395112ec.

**Comment by artur-shaik** [comment]
@The-Compiler, unfortunatly I'm quite busy right now. Not sure how long I will be busy, but if nobody will do it I can try to implement this when I will have some free time.

**Comment by The-Compiler** [comment]
Okay, thanks for the update - anyone else with a bit more JS knowledge than me stepping up? :wink:

**Comment by lahwaacz** [comment]
There are couple of things that currently don't work with webengine and are not listed above:

1. Special mouse buttons for going backward/forward in history. Works out of the box with QtWebkit and in Chromium.
2. Webengine does not understand the `qute://` scheme, so e.g. `:help` does not work.

**Comment by The-Compiler** [comment]
"Custom URL handlers" is there. I added "Other customizations in WebView/WebPage subclasses" where the mouse buttons are included.

**Comment by cosminadrianpopescu** [comment]
@The-Compiler 
I am interested in doing the WebCaret part and also the `proxy` for `QtWebEngine` if nobody is working on it. Please let me know.

**Comment by The-Compiler** [comment]
The caret part would be nice, but I think it involves a lot of javascript as the needed feature doesn't exist for QtWebEngine at all...

As for the proxy we're currently dealing with weird proxy-related crashes (see e.g. https://github.com/libproxy/libproxy/issues/45 and https://github.com/The-Compiler/qutebrowser/pull/1891#issuecomment-259265554) so I'd like to avoid adding new code for that currently.

**Comment by cosminadrianpopescu** [comment]
I don't mind the javascript part. I am ok with javascript, in general.

**Comment by The-Compiler** [comment]
I started some work on fullscreen videos in the `webengine-fullscreen` branch, but exiting fullscreen doesn't work like it should yet - I haven't found out how QupZilla handled this yet.

**Comment by Ninja3047** [comment]
They create a key event handler in webview that checks if the escape key is pressed and then it triggers the ExitFullscreen action.

```C++
void WebView::_keyReleaseEvent(QKeyEvent *event)
{
    if (mApp->plugins()->processKeyRelease(Qz::ON_WebView, this, event)) {
        event->accept();
    }

    switch (event->key()) {
    case Qt::Key_Escape:
        if (isFullScreen()) {
            triggerPageAction(QWebEnginePage::ExitFullScreen);
            event->accept();
        }
        break;

    default:
        break;
    }
}
```
https://github.com/QupZilla/qupzilla/blob/master/src/lib/webengine/webview.cpp (Line 1144)

**Comment by The-Compiler** [comment]
Many of the missing features are implemented with the upcoming v0.10.0 release ([see the changelog](https://github.com/qutebrowser/qutebrowser/blob/master/CHANGELOG.asciidoc#added)), and I think it's time to close this! 🎉 

For the things which are still missing, I opened separate issues, see [this search](https://github.com/qutebrowser/qutebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20label%3AQtWebEngine%20-label%3Abug).
